lib/eig_block_trlm.cpp

```diff
@@ -0,0 +1,771 @@
+    csParamClone.create = QUDA_ZERO_FIELD_CREATE;
+    for (int b = 0; b < block_size; b++) { r.push_back(ColorSpinorField::Create(csParamClone)); }
+
+    // Test for initial guess vectors
+    // DMH: This is an important step. With block solvers, initial guesses
+    //      of block sizes N can be subspaces rich in extremal eigenmodes,
+    //      N times more rich than non-blocked solvers.
+    //      Final paragraph, IV.B https://arxiv.org/pdf/1902.02064.pdf
+    double norm = sqrt(blas::norm2(*kSpace[0]));
+    if (norm == 0) {
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Initial residua are zero. Populating with rands.\n");
+      if (kSpace[0]->Location() == QUDA_CPU_FIELD_LOCATION) {
+        for (int b = 0; b < block_size; b++) { kSpace[b]->Source(QUDA_RANDOM_SOURCE); }
+      } else {
+        RNG *rng = new RNG(*kSpace[0], 1234);
+        rng->Init();
+        for (int b = 0; b < block_size; b++) spinorNoise(*kSpace[b], *rng, QUDA_NOISE_UNIFORM);
+        rng->Release();
+        delete rng;
+      }
+    }
+
+    bool orthed = false;
+    int k = 0, kmax = 5;
+    while (!orthed && k < kmax) {
+      orthonormalizeMGS(kSpace, block_size);
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Orthonormalising initial guesses with Gram Schmidt, k=%d\n", k);
+      orthed = orthoCheck(kSpace, block_size);
+      k++;
+    }
+
+    if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Estimate Chebyshev max.\n");
+    // Check for Chebyshev maximum estimation
+    if (eig_param->use_poly_acc && eig_param->a_max <= 0.0) {
+      // Use two vectors from kSpace as temps
+      eig_param->a_max = estimateChebyOpMax(mat, *kSpace[block_size + 1], *kSpace[block_size]);
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Chebyshev maximum estimate: %e.\n", eig_param->a_max);
+    }
+
+    // Increase evals space to nEv
+    evals.reserve(nEv);
+    for (int i = nConv; i < nEv; i++) evals.push_back(0.0);
+    //---------------------------------------------------------------------------
+
+    // Convergence and locking criteria
+    double mat_norm = 0.0;
+    double epsilon = DBL_EPSILON;
+    QudaPrecision prec = kSpace[0]->Precision();
+    switch (prec) {
+    case QUDA_DOUBLE_PRECISION:
+      epsilon = DBL_EPSILON;
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Running Eigensolver in double precision\n");
+      break;
+    case QUDA_SINGLE_PRECISION:
+      epsilon = FLT_EPSILON;
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Running Eigensolver in single precision\n");
+      break;
+    case QUDA_HALF_PRECISION:
+      epsilon = 2e-3;
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Running Eigensolver in half precision\n");
+      break;
+    case QUDA_QUARTER_PRECISION:
+      epsilon = 5e-2;
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Running Eigensolver in quarter precision\n");
+      break;
+    default: errorQuda("Invalid precision %d", prec);
```

Comment:
Can we reuse the code already in regular Lanczos here?  Perhaps create a `setEpsilon` helper in the parent class?

Reply:
I went ahead and made a bunch of helpers in the parent class. I think we could do away with having two separate classes and integrate the block solver aspects into the TRLM class.

Reply:
Let's keep them separate classes for now and just move a few helpers into the parent.  In the future where we hopefully have Arnoldi, block Arnoldi, IRLM, etc. (a girl can dream? 😉) the right level of abstraction might be different then that is now.

Reply:
On the Arnoldi, I think I know where I went wrong. The algorithm (as written in Arpack) will check to see how orthogonal the next Arnoldi vector is with respect to the current factorisation. If it is not sufficiently orthogonal, it is replaced with a random vector.

When I was developing, I was using a very small operator and the above never occurred in Arpack. Recently, I ran old faithful (L=16, T=64) through Arpack and the above happened several times. I will dig up my old Arnoldi code (which did work for small lattices) and implement this step.

Reply:
🤞 🙏 🙇‍♀️ 👍 💯